spec.bs

```diff
@@ -719,6 +719,233 @@ A [=tokenizer=] has an associated <dfn for=tokenizer>code point</dfn>, a Unicode
+
+An [=/options=] has an associated <dfn export for=options>start</dfn> property, a boolean, initially true.  If true, requires matching from the start of the input string.
+
+<h4 id=parsing>Parsing</h4>
+
+<div algorithm>
+An <dfn>encoding callback</dfn> is an abstract algorithm that takes a given string |input|.  The |input| will be a simple text piece of a [=/pattern string=].  An implementing algorithm should validate and encode the |input|.  It must return the encoded string or throw an exception.
+</div>
+
+A <dfn>pattern parser</dfn> is a [=struct=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">token list</dfn>, a [=/token list=], initially an empty [=list=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">encoding callback</dfn>, a [=/encoding callback=], initially null.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">options</dfn>, a [=/options=], initially null.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">segment wildcard regexp</dfn>, a string, initially the empty string.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">part list</dfn>, a [=/part list=], initially an empty [=list=].
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">pending fixed value</dfn>, a string, initially the empty string.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">index</dfn>, a number, initially 0.
+
+A [=pattern parser=] has an associated <dfn for="pattern parser">next numeric name</dfn>, a number, initially 0.
+
+The <dfn>full wildcard regexp value</dfn> is the string "`.*`".
+
+<div algorithm>
+To <dfn>try to consume a token</dfn> given a [=pattern parser=] |parser| and [=token/type=] |type|:
+
+  1. Let |next token| be |parser|'s [=pattern parser/token list=][|parser|'s [=pattern parser/index=]].
+  1. If |next token|'s [=token/type=] is not |type| return null.
+  1. Increment |parser|'s [=pattern parser/index=] by 1.
+  1. Return |next token|.
+</div>
+
+<div algorithm>
+To <dfn>try to consume a modifier token</dfn> given a [=pattern parser=] |parser|:
+
+  1. Let |token| be the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`other-modifier`</a>".
+  1. If |token| is not null, then return |token|.
+  1. Set |token| to the result of running [=try to consume a token=] given |parser| and "<a for=token/type>`asterisk`</a>".
+  1. Return |token|.
+</div>
+
+<div algorithm>
+To <dfn>consume a required token</dfn> given a [=pattern parser=] |parser| and [=token/type=] |type|:
+
+  1. Let |result| be the result of running [=try to consume a token=] given |parser| and |type|.
+  1. If |result| is null, then throw a {{TypeError}}.
+  1. Return |result|.
+</div>
+
+<div algorithm>
+To <dfn>consume text</dfn> given a [=pattern parser=] |parser|:
+
+  1. Let |result| be the empty string.
+  1. Let |done| be false.
+  1. While |done| is false:
```

Comment:
Can you avoid the done variable by just using `[=iteration/break=]` and "While true:"?

Reply:
I really want a `do {} while()` loop.  What is the best spec construct for that?

Reply:
I don't understand why a do/while would be necessary here. If the first token ends up being null, you do nothing, right? So just a normal while loop would work.

Reply:
I don't think that's the case?

https://source.chromium.org/chromium/chromium/src/+/main:third_party/liburlpattern/parse.cc;l=79-87;drc=ccecd870311e5326a355715201778816c841e8f0

I guess I can just duplicate the two "try to consume" steps to get the token before the loop.  Just using `While true:` is probably easier, though.

Reply:
I'm still confused. Here is my proposal:

While true:

1. Let token be the result of running try to consume a token given parser and "char".
1. If token is null, then set token to the result of running try to consume a token given parser and "escaped-char".
1. If token is null, then break.
1. Append token's value to the end of result.

Let's say both "char" and "escaped-char" will return null for the first token. Then:

- in your version we run steps 1, 2 which have side effects. 3 sets done to true. 4 does nothing since it's "else". Then we go back to the top of the loop where done is false, so we're done.

- in my version we run steps 1, 2. 3 breaks. So we're done.

Reply:
(I guess it's not clear whether there are actually side effects... it seems like the index doesn't advance in that case.)

Reply:
Yea, your proposal works.  Its equivalent to the "do while" without the break statement.